File: apps/web/src/lib/models/fluids.ts
/**
 * Fluid definition and properties models.
 * Mirrors: apps/api/src/opensolve_pipe/models/fluids.py
 */

/** Available fluid types. */
export type FluidType =
	| 'water'
	| 'ethylene_glycol'
	| 'propylene_glycol'
	| 'diesel'
	| 'gasoline'
	| 'kerosene'
	| 'hydraulic_oil'
	| 'custom';

/** Display names for fluid types. */
export const FLUID_TYPE_LABELS: Record<FluidType, string> = {
	water: 'Water',
	ethylene_glycol: 'Ethylene Glycol',
	propylene_glycol: 'Propylene Glycol',
	diesel: 'Diesel',
	gasoline: 'Gasoline',
	kerosene: 'Kerosene',
	hydraulic_oil: 'Hydraulic Oil',
	custom: 'Custom Fluid'
};

/** Fluid types that require concentration. */
export const GLYCOL_FLUID_TYPES: FluidType[] = ['ethylene_glycol', 'propylene_glycol'];

/** Definition of the working fluid. */
export interface FluidDefinition {
	/** Type of fluid. Default: 'water' */
	type: FluidType;
	/** Operating temperature in project units (F or C). Default: 68.0 */
	temperature: number;
	/** Concentration percentage for glycol mixtures (0-100). */
	concentration?: number;
	/** Custom fluid density (kg/m³ for SI calculation). Required if type is 'custom'. */
	custom_density?: number;
	/** Custom fluid kinematic viscosity (m²/s for SI calculation). Required if type is 'custom'. */
	custom_kinematic_viscosity?: number;
	/** Custom fluid vapor pressure (Pa for SI calculation). Required if type is 'custom'. */
	custom_vapor_pressure?: number;
}

/** Calculated fluid properties at operating conditions (always in SI units). */
export interface FluidProperties {
	/** Density in kg/m³. */
	density: number;
	/** Kinematic viscosity in m²/s. */
	kinematic_viscosity: number;
	/** Dynamic viscosity in Pa·s. */
	dynamic_viscosity: number;
	/** Vapor pressure in Pa. */
	vapor_pressure: number;
	/** Specific gravity relative to water at 4°C. Default: 1.0 */
	specific_gravity: number;
}

/** Default fluid definition (water at 68°F). */
export const DEFAULT_FLUID_DEFINITION: FluidDefinition = {
	type: 'water',
	temperature: 68.0
};

/** Check if a fluid type requires concentration. */
export function requiresConcentration(type: FluidType): boolean {
	return GLYCOL_FLUID_TYPES.includes(type);
}

/** Check if a fluid type is custom (requires all custom properties). */
export function isCustomFluid(type: FluidType): boolean {
	return type === 'custom';
}

/** Validate fluid definition. Returns error message or null if valid. */
export function validateFluidDefinition(fluid: FluidDefinition): string | null {
	if (requiresConcentration(fluid.type) && fluid.concentration === undefined) {
		return `${FLUID_TYPE_LABELS[fluid.type]} requires concentration to be specified`;
	}

	if (isCustomFluid(fluid.type)) {
		const missing: string[] = [];
		if (fluid.custom_density === undefined) missing.push('custom_density');
		if (fluid.custom_kinematic_viscosity === undefined) missing.push('custom_kinematic_viscosity');
		if (fluid.custom_vapor_pressure === undefined) missing.push('custom_vapor_pressure');
		if (missing.length > 0) {
			return `Custom fluid requires: ${missing.join(', ')}`;
		}
	}

	if (fluid.concentration !== undefined && (fluid.concentration < 0 || fluid.concentration > 100)) {
		return 'Concentration must be between 0 and 100';
	}

	return null;
}

File: apps/api/src/opensolve_pipe/models/fluids.py
"""Fluid definition and properties models."""

from enum import Enum

from pydantic import Field, model_validator

from .base import NonNegativeFloat, OpenSolvePipeBaseModel, PositiveFloat


class FluidType(str, Enum):
    """Available fluid types."""

    WATER = "water"
    ETHYLENE_GLYCOL = "ethylene_glycol"
    PROPYLENE_GLYCOL = "propylene_glycol"
    DIESEL = "diesel"
    GASOLINE = "gasoline"
    KEROSENE = "kerosene"
    HYDRAULIC_OIL = "hydraulic_oil"
    CUSTOM = "custom"


class FluidDefinition(OpenSolvePipeBaseModel):
    """Definition of the working fluid."""

    type: FluidType = Field(default=FluidType.WATER, description="Type of fluid")
    temperature: float = Field(
        default=68.0, description="Operating temperature in project units (F or C)"
    )
    concentration: float | None = Field(
        default=None,
        ge=0,
        le=100,
        description="Concentration percentage for glycol mixtures",
    )

    # For custom fluids - all required if type is CUSTOM
    custom_density: PositiveFloat | None = Field(
        default=None, description="Custom fluid density (kg/m³ for SI calculation)"
    )
    custom_kinematic_viscosity: PositiveFloat | None = Field(
        default=None,
        description="Custom fluid kinematic viscosity (m²/s for SI calculation)",
    )
    custom_vapor_pressure: NonNegativeFloat | None = Field(
        default=None, description="Custom fluid vapor pressure (Pa for SI calculation)"
    )

    @model_validator(mode="after")
    def validate_custom_fluid(self) -> "FluidDefinition":
        """Validate that custom fluids have all required properties."""
        if self.type == FluidType.CUSTOM:
            missing = []
            if self.custom_density is None:
                missing.append("custom_density")
            if self.custom_kinematic_viscosity is None:
                missing.append("custom_kinematic_viscosity")
            if self.custom_vapor_pressure is None:
                missing.append("custom_vapor_pressure")
            if missing:
                raise ValueError(
                    f"Custom fluid requires: {', '.join(missing)}"
                )
        return self

    @model_validator(mode="after")
    def validate_glycol_concentration(self) -> "FluidDefinition":
        """Validate that glycol fluids have concentration specified."""
        if (
            self.type in (FluidType.ETHYLENE_GLYCOL, FluidType.PROPYLENE_GLYCOL)
            and self.concentration is None
        ):
            raise ValueError(
                f"{self.type.value} requires concentration to be specified"
            )
        return self


class FluidProperties(OpenSolvePipeBaseModel):
    """Calculated fluid properties at operating conditions (always in SI units)."""

    density: PositiveFloat = Field(description="Density in kg/m³")
    kinematic_viscosity: PositiveFloat = Field(
        description="Kinematic viscosity in m²/s"
    )
    dynamic_viscosity: PositiveFloat = Field(description="Dynamic viscosity in Pa·s")
    vapor_pressure: NonNegativeFloat = Field(description="Vapor pressure in Pa")
    specific_gravity: PositiveFloat = Field(
        default=1.0, description="Specific gravity relative to water at 4°C"
    )
